### ISSUE TITLE
after nginx, and the domain  not send to client

### ISSUE THREAD

**Issue by coder-free** [boringproxy/boringproxy#127]
Hi, I running boringproxy after nginx. the local server can not get the domain.
Specifically:
VPS:
nginx start on port 80 and 443. running with https.
boringproxy running as server on port 18888. running with http.

local server:
nginx start on port 80. running with http.
and nginx reverse proxy by domain. nginx.conf like this:
```
### some thing ...
http {
    ### some thing ...
    # default server
    server {
        listen       80;
        server_name  localhost;

        location / {
            root   html;
            index  index.html index.htm;
        }
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   html;
        }
    }

    # server1
    server {
        listen       80;
        server_name  my.domain.com;

        location / {
            root   myhtml;
            index  index.html index.htm;
        }
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   myhtml;
        }
    }

    # server2
    server {
        listen       80;
        server_name  my2.domain.com;

        location / {
            root   my2html;
            index  index.html index.htm;
        }
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   my2html;
        }
    }
}
```
Then no matter which domain name request is used, it will enter the default server.

When I and tunnel, the "TLS Termination" I selected "Server".

If the "TLS Termination" I selected "Client". My boringproxy client(running on my local server) notice me request https certificate. The message is:
```
Your sites will be served over HTTPS automatically using an automated CA.
By continuing, you agree to the CA's terms of service.
Please enter your email address to signify agreement and to be notified
in case of issues. You can leave it blank, but we don't recommend it.
  Email address:
```

When I and tunnel, the "TLS Termination" I selected "Passthrough".

In summary:
The "TLS Termination" I selected "Passthrough" or "Server". My local server nginx can not reverse proxy by domain.
The "TLS Termination" I selected "Client". Boringproxy client said need https certificate.

I think the request path:
```
browser ___https____> nginx(on vps) _____reverse proxy_____> boringproxy(as server on vps) ____some protocol____> boringproxy(as client on local server) ___http___> nginx(on local server)
```
is safe.

Just make sure that "browser to nginx(on vps) " and "boringproxy(as server) to boringproxy(as client)" are safe. The request will be safe.

Maybe mine is a strange demand 🤣.

**Comment by anderspitman** [comment]
Ok I think I understand what you're trying to do. I feel like your setup needs to be simplified, but this is also a good test to potentially fix some bugs in boringproxy, so let's see if we can make it work. Some followup questions:

Given that you're terminating TLS at the nginx server, you definitely don't want to use Client TLS Termination. The client isn't yet updated to support disabling auto certs (added #129 to track that). Passthrough might work, but I would recommend Server TLS Termination.

Are you sure that your server nginx is passing the X-Forwarded-Host header?

**Comment by anderspitman** [comment]
I think it looks like this:

```
location / {
    proxy_set_header X-Forwarded-Host  $host;
}
```

**Comment by coder-free** [comment]
My nginx config is:

```
location / {
    client_max_body_size 0;
    gzip off;

    proxy_read_timeout      300;
    proxy_connect_timeout   300;
    proxy_redirect          off;

    proxy_http_version 1.1;
    proxy_set_header    Host                $http_host;
    proxy_set_header    X-Real-IP           $remote_addr;
    proxy_set_header    X-Forwarded-Ssl     on;
    proxy_set_header    X-Forwarded-For     $proxy_add_x_forwarded_for;
    proxy_set_header    X-Forwarded-Proto   $scheme;
    proxy_pass http://my-workhorse;
}
```

And I change it to:
```
location / {
    client_max_body_size 0;
    gzip off;

    proxy_read_timeout      300;
    proxy_connect_timeout   300;
    proxy_redirect          off;

    proxy_http_version 1.1;
    proxy_set_header    Host                $http_host;
    proxy_set_header    X-Forwarded-Host    $host;
    proxy_set_header    X-Real-IP           $remote_addr;
    proxy_set_header    X-Forwarded-Ssl     on;
    proxy_set_header    X-Forwarded-For     $proxy_add_x_forwarded_for;
    proxy_set_header    X-Forwarded-Proto   $scheme;
    proxy_pass http://my-workhorse;
}
```

The result has not changed.

**Comment by anderspitman** [comment]
Don't have time at the moment to test myself, but this might be related to #62.

**Comment by coder-free** [comment]
@anderspitman Has there been an update on this issue recently?
I don't know if the tunnel can send the host header. I'm wondering if the boringproxy client can send a request with a path, so that the client's nginx can also use the path to forward different requests.
```
browser -> nginx(on vps) -> boringproxy(as server on vps) -> 
boringproxy(as client on local server) ___(with additional path)___> nginx(on local server)
```
maybe like this:
![image](https://user-images.githubusercontent.com/2630290/151159924-3e281c0d-aee5-434b-9071-44c55d89c817.png)

if an url is :
```
http://aaa.com/api/user/login
```
then the nginx(on local server) received url is:
```
http://aaa.com/test/api/user/login
```
and the nginx config file use rewrite module to regress the url, may like this:
```
location /test-work {
    client_max_body_size 0;
    gzip off;

    proxy_read_timeout      300;
    proxy_connect_timeout   300;
    proxy_redirect          off;

    proxy_http_version 1.1;
    proxy_set_header    Host                $http_host;
    proxy_set_header    X-Real-IP           $remote_addr;
    proxy_set_header    X-Forwarded-Ssl     on;
    proxy_set_header    X-Forwarded-For     $proxy_add_x_forwarded_for;
    proxy_set_header    X-Forwarded-Proto   $scheme;

    rewrite /test / break;

    proxy_pass http://test-work;
}
```

On my local server, there running two or more service, I need forward requests to different service.

**Comment by almereyda** [comment]
When forwarding to multiple services, you can use boringproxy as your reverse proxy and point multiple (sub-)domains to the different upstream servers directly.

It feels like path-based service routing is deprecated in contemporary web service deployments, where routing different services via a subdomain is preferred. Also many services expect to be run at the `/` of a domain, and do not support sitting behind a path-based (sub-)routing themselves.

1. Have you considered deploying your services independently, without an intermediary reverse proxy? This way you could proxy to their listening ports directly.
2. Else, are there explicit constraints that you need to meet, which make you favour the one option over the other?

To respond on your example, it seems it is a little complicated, where you introduce another reverse proxy in front of the boringproxy, while wanting to terminate the TLS connections on the clients, only after a request already passed two stations, followed by another reverse proxy. I suggest to look into ways that allow you to build a tunnel in the form of:

- `client > boringserver > boringclient > service`

This is also why boringproxy originally requests both ports 80 and 443 on the server it runs, in so it gets full control over the HTTP connections. If you need to use it as a forwarding backend service itself, like you explained, I'd wonder if you could omit any TLS termination through it at all, and do everything in your primary nginx entrypoint.

Btw., Caddy is also a nice webserver, that has a simpler configuration syntax than nginx, and most of the time just works (TM).

If you could omit the requirement to have the TLS connection terminated at the client, you would probably also not run into #62

**Comment by anderspitman** [comment]
Thanks @almereyda. I agree with you and would really prefer not to add path-based routing to boringproxy if possible. For one thing there are security implications with cookies being shared between different applications on the same domain. @coder-free do you think you can avoid paths?

**Comment by coder-free** [comment]
@anderspitman @almereyda Paths are not required. I just want a way to distribute my requests to different services on my local server.
I have little internet knowledge. I don't know why the domain not send to client. I'm not sure if it's an nginx problem or a boringproxy problem. Now my request path is `client > nginx(1) > boringserver > boringclient >  nginx(2) > service`. The nginx(1) is required. My vps running some other service. I now distinguish different services through different domain names. For example:
`www.mydomain.com` will be forwarded to my home page.
`git.mydomain.com` will be forwarded to the gitlab service.
`nexus.mydomain.com` will be forwarded to the maven service.
and etc.
Gitlab and nexus need more memory. My VPS have not enough memory. So I want put gitlab and nexus on my local server. But different domain names are not recognized on the local server.

**Comment by anderspitman** [comment]
Ok in theory this should be working, but in terms of just getting you up and running, is there some reason you can't drop nginx(2)? Other than performance things like caching and compression, boringproxy should have most of the features necessary to act as a reverse proxy in place of nginx.

So it would be:

`client > nginx(1) > boringserver > boringclient > service`

You just need to create a separate tunnel for each domain in the boringproxy UI.

**Comment by anderspitman** [comment]
If your services are on the same machine, you can proxy them both through a single boringproxy client. If they are running on separate machines, you can run multiple instances of the client, one on each machine.

**Comment by coder-free** [comment]
Oh, this is my mistake. I thought that a client can only correspond to one tunnel.
Now I don't need to remove nginx2. I just need to make gitlab work on port 81 and nexus on port 82, and then create two tunnels to point to these two ports respectively.
Additional:
The simple proxy to gitlab is proxy to a unix socket. Nginx config like this:
```
    upstream gitlab-workhorse {
        server unix:/var/opt/gitlab/gitlab-workhorse/sockets/socket fail_timeout=0;
    }
    server {
        listen 0.0.0.0:81;
        listen [::]:81 ipv6only=on default_server;
        
        ...

        location / {

            ...

            proxy_pass http://gitlab-workhorse;
        }
    }
```
If I remove nginx2. I need to reconfig gitlab to make it work on a certain port. But that doesn't matter now. No need to remove nginx2, boringproxy does all my needs.
